Aplib.Core/ExposedQueue.cs

```diff
@@ -0,0 +1,192 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// A queue with all elements exposed.
+    /// Functionally works like a queue with indexing.
+    /// </summary>
+    public class ExposedQueue<T> : ICollection<T>
+    {
+        /// <summary>
+        /// The length of the array.
+        /// </summary>
+        public int MaxCount { get; private set; }
+
+        /// <summary>
+        /// Actual number of elements in the array.
+        /// </summary>
+        public int Count { get; private set; }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly => false;
+
+        private readonly T[] _array;
+        private int _head;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// </summary>
+        /// <param name="size">The size of the array.</param>
+        public ExposedQueue(int size)
+        {
+            MaxCount = size;
+            Count = 0;
+            _array = new T[MaxCount];
+            _head = MaxCount - 1;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// By default, assumes the array is filled.
+        /// </summary>
+        /// <param name="array">An array to use as the circular array.</param>
+        /// <param name="count">The number of actual elements in the array.</param>
+        public ExposedQueue(T[] array, int? count = null)
+        {
+            if (count > array.Length)
+                throw new ArgumentOutOfRangeException(nameof(count), "Count cannot exceed the length of the array.");
```

Comment:
Shouldn't you be allowed to initialize the queue with an array that is less than the maximal elements?

Furthermore, if the count is less than the length of the array, why are we still storing the whole array? Wouldn't it be better if we copied the array to another array with the length of `count`.